nncf/auto/environment/quantization_env.py

```diff
@@ -0,0 +1,669 @@
+        # Quantizer Master Table Creation
+        # TODO: Discuss how to organize the following lines (they are dependent sequentially)
+        self._groups_of_adjacent_quantizers = GroupsOfAdjacentQuantizers(self.qctrl)
+        self.learnable_qids, self.lead_qid_of_groups = self._get_learnable_qids_and_leads()
+        self.quantizer_table = self._create_quantizer_table()
+
+        # Create master dataframe to keep track of quantizable layers and thier attributes, a.k.a state embedding
+        self.master_df, self.state_list = self._get_state_space(self.qctrl, self.qmodel, self.quantizer_table)
+    
+        assert len(self.quantizer_table) == len(self.qctrl.all_quantizations), \
+            "Number of Quantizer is not tally between quantizer table and quantization controller"
+        
+        # MinMaxScaler for State Embedding
+        self.state_scaler = MinMaxScaler()
+        self.state_scaler.fit(self.master_df[self.state_list])
+
+        # Model Size Calculation
+        self.orig_model_size   = sum(self.master_df['param']*self.master_df.is_wt_quantizer)*self.float_bitwidth  #in bit unit
+        self.min_model_size    = sum(self.master_df['param']*self.master_df.is_wt_quantizer)*self.min_bitwidth    # This variable has only been used once, just to ensure size constrainer doesnt go below than this
+        self.target_model_size = self.orig_model_size*self.compression_ratio
+
+        # Evaluate and store metric score of pretrained model
+        self._evaluate_pretrained_model()
+
+        # init reward
+        self.best_reward = -math.inf #TODO: move reward to search manager
+        self.reset()
+        
+        # Serialize Q.Env information. Note that these functions should be at the end of Q.Env Initialization.
+        self._dump_master_df()
+        self._dump_quantized_graph()
+        self._dump_groups_of_adjacent_quantizers()
+
+        # End of QuantizationEnv.__init__()
+        # ----------------------------------------------------------------------------------------------------------------------
+
+    def _get_learnable_qids_and_leads(self):
+        # By default, all weight quantizers must be learnable to allow optimal model size compression
+        # Bitwidth assignment mode only applies to activation quantizers of a group. 
+        # When the mode is LIBERAL, all activation quantizers are learnable.
+        # When it is STRICT, none of the activation quantizers is learnable but to follow the leader of the group.
+        # HW config determines the bitwidth space of a quantizer
+        # TODO: discuss on STRICT_SOFT and STRICT_HARD
+
+        learnable_qids = []
+        lead_qid_of_groups = [None]*len(list(self._groups_of_adjacent_quantizers))
+
+        if self.bw_assignment_mode is BitwidthAssignmentMode.STRICT:
+
+            for i, g in enumerate(self._groups_of_adjacent_quantizers):
+                n_wq = len(self._groups_of_adjacent_quantizers._groups_of_adjacent_quantizers[i].weight_quantizers)
```

Comment:
feel free to extend `GroupsOfAdjacentQuantizers` class by `__getitem__` for the direct access to the group. It would remove unnecessary noise from your code:
```
n_aq = len(self._groups_of_adjacent_quantizers[i].activation_quantizers)
```

Reply:
Bitwidth assignment for AutoQ is decided to be liberal at all cases. This routine has been removed.